Create eip-3778 (#3778)
ethereum/EIPs: EIPS/eip-3777.md

---

> eip: 3777

Reviewer: ```suggestion
eip: 3778
```

title: Time proof non-fungible Token

> author: Cyberforker(cyberforker@outlook.com)

Reviewer: ```suggestion
author: Cyberforker (@CyberFork)

```

discussions-to: https://ethereum-magicians.org/t/a-time-proof-non-fungible-token/

> status: draft

Reviewer: ```suggestion
status: Draft

```

type: Standards Track
category: ERC
created: 2021-09-01
requires: 165, 721
---

> ## Simple Summary

Reviewer: ```suggestion
## Simple Summary
```
This section is no longer valid, please use the `description` preamble field. (see the `eip-template.md` for more info)

This is a standard for time-stamp provable NFTs, which extends [ERC-721](./eip-721.md).
Enables NFT to mapping and anchor things that have time features. 
And also the time-stamp proof of NFTs can be the condition of interaction with smart contracts.
It is called Time proof Non-fungible Token and is referred to as tpNFT in the subsequent descriptions.
## Abstract

> Based on the tokenURI parameter of ERC721, the plain text parameter and the encryption parameter are added, and the time proof is set for the update of these three parameters.

Reviewer: ```suggestion
Based on the `tokenURI` parameter of ERC-721, the plain text parameter and the encryption parameter are added, and the time proof is set for the update of these three parameters.

```

- Scientific discoveries: Publish the publicly available content of relevant scientific discoveries in plain text, and fix some temporarily undisclosed demonstration details through encrypted parameters. Engrave the brilliance of human wisdom on the blockchain. The timestamp proves that it will not lie.
- Prophecy: Make predictions and cast them into tpNFT with time proof ability, as proof of personal ability, or as a certificate for receiving bonuses.
- Identity: Participating in a series of on-chain behaviors through tpNFT, including the prophecy mentioned in the previous point, revolves around tpNFT's timestamp proof, which has become a natural proof of identity.
- Legacy: Binding assets to a tpNFT, others can receive part or all of the assets by submitting the preimage of encryption parameter(Hash or HashTree). We have another EIP to solve the front-running attacks.
- Dynamic NFT: With tpNFT, we only need a little improvement then can get a brand new type of NFT which state on the chain will automatically change over time without consuming any gas fees. Such features can open up new areas for games, installation art, financial contracts, and so on.
   

> Summarize: From a philosophical point of view, The NFT of the ERC721 standard is mainly used to anchor and map things related to space, while tpNFT is used for related to time.

Reviewer: ```suggestion
Summarize: From a philosophical point of view, The NFT of the ERC-721 standard is mainly used to anchor and map things related to space, while tpNFT is used for related to time.

```

    4. Gas savings design
       1. The uint64 of four timestamps variables in the struct of TimeProof are arranged consecutively in the struct as one uint256.
       2. The maximum value of Uint64 is 18446744073709551615 (584942417355 years), which is sufficient for saving the timestamp.  
       3. In three update functions, if the value of the calldata and the parameter(URI or description or hashedproof) in the storage of tpNFT are both zero, the update will be skipped to save Gas and meet the needs of mint empty NFT.
   
**ERC-165 Interface**
We chose Standard Interface Detection (ERC-165) to expose the interfaces that a tpNFT smart contract supports.

## Backwards Compatibility
 1. Compatible with ERC165 and the implementation of the relevant interface and registration.
 2. Based on ERC721 and fully compatible, it is a superset of ERC721. If you only fill in the tokenURI and do not fill in the description and hash in the mint, and then set tpNFT to be constant, you will get a tpNFT that only has more tokenURI update timestamps than the standard NFT with tokenURI.
## Test Cases

> ### Mint

Reviewer: Please remove these empty sections.

    2. When updating the data of tpNFT, it is prohibited to directly modify the TimeProof struct in the _timeProofs mapping, must use the three internal methods _updateURI, _updateDescription, and _updateHashProof provided by the standard to update the data of tpNFT so that the updated content and the updated timestamp are in the same atom transaction and executed correctly.
3. Keccak256 is selected as the function of Hash, it is widely used and its safety has been tested for a long time.
    1. Anti-collision: Can not find two different data with the same Hash value.
    2. Irreversible: Raw raw data cannot be obtained by reverse calculation of Hash value.
    3. Leakage risk: We provide a simple function [genHashProof] to pass in the string and generate the hash value. It should be clarified that this function is for reference only. The preimage may be leaked due to RPC and other reasons when using this method of deploying the contract on the chain. Therefore for safety, if your preimage is very important, it is recommended that you copy the source code to the local offline or use other SDK to generates hash values ​​in a local offline environment.

> ## References

Reviewer: This is not a valid EIP section, please remove.

**Standards**

1. [ERC-20](./eip-20.md) Token Standard.
2. [ERC-165](./eip-165.md) Standard Interface Detection.
3. [ERC-721](./eip-721.md) Non-Fungible Token Standard.
## Copyright

> Copyright and related rights waived via CC0.

Reviewer: ```suggestion
Copyright and related rights waived via [CC0](https://creativecommons.org/publicdomain/zero/1.0/).
```

Author: Thanks for guiding the revision, it has been corrected, please check again.
